includes/backend.inc

```diff
@@ -1173,7 +1173,16 @@ function _drush_backend_generate_command($site_record, $command, $args = [], $co
       $ssh_cmd[] = '-t';
     }
     $ssh_cmd[] = $username . drush_escapeshellarg($hostname, "LOCAL");
-    $ssh_cmd[] = drush_escapeshellarg($command . ' 2>&1', "LOCAL");
+
+    // Escape the remote command for the LOCAL OS, then add in the proper quote style based on the host OS
+    $remote_cmd = drush_escapeshellarg($command . ' 2>&1', "LOCAL", TRUE);
+    if (drush_is_windows()) {
+      $remote_cmd = '"' . $remote_cmd . '"';
+    }
+    else {
+      $remote_cmd = "'" . $remote_cmd . "'";
```

Comment:
I don't think that you can reliably just wrap an escaped command in single-quotes & trust that it will correctly escape an escaped command,  e.g. if $remote_cmd contains a single quote.

Reply:
$remote_cmd is still escaped two lines above that with $raw = TRUE.  This is just doing what the escape functions do when $raw is FALSE, while still escaping for LOCAL.

Reply:
If you escape `$foo` to `'$foo'`, then wrapping that in single quotes will give you `''$foo''`, which is not correct.

Reply:
I agree, that would be bad.

With this code change, the call to `drush_escapeshellarg($arg, $os, $raw)` (line 1178) has `$raw` set to `TRUE`.  Both the `_drush_escapeshellarg_linux($arg, $raw=FALSE)` and `_drush_escapeshellarg_windows($arg, $raw=FALSE)` functions, which are used by this parent function, are passed `$raw`.  From what I'm seeing, having `$raw = TRUE`  ensures `$foo` remains `$foo` after line 1178 has completed execution.

If `'$foo'` is passed as $arg (single-quotes included) on line 1178, then the Linux escape will make it `'\''$foo'\''`, as you mentioned, and then this PR will make it `''\''$foo'\'''`.  Similar-ish things will happen with Windows and double-quotes.  Is this the edge-case you're talking about?  That would imply drush is being called in a manner similar to: `drush @remote-host 'command'`.

Is this a supported use case?  If so, would stripping existing wrapped quotes from `$arg` before escaping take care of this concern?

Reply:
Looking at the `$raw` handling, it seems that your code block above passes `$raw` as True to avoid wrapping the result in `"` or `'`, and then immediately puts back on the quote characters that it asked to be omitted.

I don't remember why we have the `$raw` parameter at all. It seems that in the case of the code above, you should let `$raw` stay FALSE. Then you should get the result you expect without having to manually re-add the quotes yourself.

Note that it's still not entirely clear to me which exact character escape combinations this PR is intended to fix. The existing escaping code appears to work really well for Linux-to-Linux commands, so it would be a good idea to ensure that code path passes data through the same way it does today. It would also be really helpful if you could add failing test cases (that is, tests that fail without this PR, and pass with it).

Reply:
This is a Windows-to-Linux issue, something I'm guessing the tests don't cover.  I discovered all of this when using the Windows command line and trying to use an alias for a remote server at Acquia.

The original escape happens for the LOCAL, or target, OS.  The escape function assumes that if `$os` doesn't start with "WIN" that it should escape `$arg` as Linux.  Since line 1178 has `$os = LOCAL` (matching the original code), the whole command is escaped as Linux.  With `$raw = TRUE`, we need to add wrapping quotes back in, which is what the rest of the PR does.

With (Open)SSH on Windows command line/PowerShell, the SSH call in `proc_open()` fails if single quotes are used.  However, the same SSH call works when double-quotes are used.  Setting `$raw = TRUE` and wrapping with quotes based on the host OS is what this PR is trying to address.

Ultimately, the problem seems to be the use of `$os = LOCAL` when escaping the SSH command.  If the escaping only needs to happen for the host machine, then `$os` shouldn't be set and this PR would simply change the one-line function call to exclude LOCAL.  If the command needs to be escaped for the LOCAL OS, which the existing code does, then this PR takes care of escaping the SSH command for the LOCAL OS and wrapping the command with quotes that are appropriate for the host OS.  This way the SSH command is escaped for the LOCAL OS, but wrapped with quotes based on the host OS.

Reply:
I think that perhaps the code needs to do what you said in the issue description -- first call `drush_escapeshellarg` for the target platform, and then, if the local platform is different, call `drush_escapeshellarg` for the local platform.

I think that setting `$raw` and manually wrapping in quotes isn't going to work. Make sure that you can escape complex commandline arguments, e.g.
```
drush ev '$item = menu_get_item(); var_export($item);'
```
> The escape function assumes that if $os doesn't start with "WIN" that it should 
> escape $arg as Linux.

That sounds like a bug. If $os is "LOCAL", then the local OS should be looked up and used to determine the escape mechanism.

Reply:
I've updated the PR to better align with convention.

I'm not entirely sure how "LOCAL" is meant to be used, but `_drush_get_os()` makes it sound like it's meant to be used as documentation-by-code.  However, `drush_escapeshellarg()` doesn't take "LOCAL" into account in any way.  One of my iterations on this locally was to place `$os = strtoupper($os) == 'LOCAL' ? NULL : $os;` at the top of this function, but I decided that sort of a change would have a far wider impact than what this PR does.